### PR TITLE
[IB method] Implemented new ghost point method

### PR DESCRIPTION
**Pull request by ktoepperwien**
Implemented new ghost point method following [Ghias et al., JCP 225(1), 2007](https://linkinghub.elsevier.com/retrieve/pii/S0021999106006048), with image point interpolation based on [R. Franke, Mathematics of Computation 38(157), 1982](https://www.ams.org/mcom/1982-38-157/S0025-5718-1982-0637296-4/). Image points currently have the same distance from IB surface as ghost points. In this commit, only Dirichlet boundary conditions supported! Includes test case for non-reacting flow over 2D hill ("Witch of Agnesi"/"Agnesi hill") based on [Lundquist et al., Month. Meterol. Rev. 140(12), 2012](http://journals.ametsoc.org/doi/10.1175/MWR-D-11-00311.1).

Notes:
- `common_ops.py` scatter/gather is incompatible with lists, therefore commented out dim-check.
- To run Witch of Agnesi test case, be sure to initialize the Coriolis force in `fire.py` as follows: 
   ```_DEFAULT_LAT = 0.5497607357 # rad == 31.5 deg 
   # Coriolis force coeff of 1e-4 rad/s from Lundquist et al., Month. Meterol.
   # Rev. 2010 for validation of IB methods corresponds to a latitude 
   # defined by _REF_LAT 
   _REF_LAT = 0.7555285998 # rad == 43.289 deg 
   my_lat = _REF_LAT 
   logging.warn(f'Using reference latitude of {my_lat} rad ' 
                f'corresonding to {np.rad2deg(my_lat):.3f} deg.') 
   self.coriolis_force_fn = cloud_utils.coriolis_force(my_lat, { 
       'u': self.fire_utils.u_init, 
       'v': self.fire_utils.v_init, 
       'w': 0.0
   }, 2)```

### CODE REVIEW

**Comment by john-qingwang** [comment]
There are quite a bit of lint issues that are not conforming with the Google programming style. I can go ahead and correct them, but I think it would be a good practice for you to work on it. You may found https://github.com/google/yapf useful for this purpose. Thanks!

**Comment by ktoepperwien** [comment]
Note: To preserve second order accuracy with Neumann boundary conditions, we need 2 additional points in the fluid domain, see `Eq. 14` in Luo et al., International Journal of Heat and Mass Transfer 92, 2016 ([Link](https://linkinghub.elsevier.com/retrieve/pii/S0017931015009679), [DOI: 10.1016/j.ijheatmasstransfer.2015.09.024](https://doi.org/10.1016/j.ijheatmasstransfer.2015.09.024)). This is not yet implemented -- we are currently using a single image point for Neumann boundary conditions.
For non-isothermal cases, we will also need to update the density at the ghost point by invoking the equation of state (which is only first-order accurate). For second-order accuracy, see `Eqs. 17` and `18` in above reference.

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 1]
```diff
@@ -1124,6 +1699,252 @@ def _apply_feedback_force_1d_interp(
 
     return ib_force
 
+  def _ib_gp_force_fn(
+      self,
+      interp_weights: FlowFieldVal,
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:1704]: The selection of these weights depends on the type of BC we'd like to enforce. One of them is redundant regardless of the type of boundary conditions. Let's make it an argument of the returned function to eliminate the ambiguity.

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 2]
```diff
@@ -1124,6 +1699,252 @@ def _apply_feedback_force_1d_interp(
...
+      # Since the ghost point values are stored in a 1D-vector, we need to
+      # scatter these values to the corresponding grid node in the 3D grid,
+      # i.e., generate a 3D tensor.
+      # gp_tensor will only have non-zero values at ghost point indices
+      # (ijk_gp), all other points will be zero.
+      gp_tensor = common_ops.scatter(
+          gp_val,
+          ijk_gp,
+          shape=[
+              self._params.core_nz + 2 * self._params.halo_width,
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:1840]: In the context of Swirl-LM, `params.nx` is the number of point along the x direction local to a core, including halos. It can be used directly here.

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 3]
```diff
@@ -1124,6 +1699,252 @@ def _apply_feedback_force_1d_interp(
...
+              self._params.core_nx + 2 * self._params.halo_width,
+              self._params.core_ny + 2 * self._params.halo_width
+          ],
+          dtype=tf.float32
+      )
+
+      return tf.nest.map_structure(
+          lambda _u, _u_F: -beta * (_u - _u_F), u, gp_tensor
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:1848]: Let's follow the naming rules to set all variable names with lower case letters.

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 4]
```diff
@@ -1124,6 +1699,252 @@ def _apply_feedback_force_1d_interp(
 
     return ib_force
 
+  def _ib_gp_force_fn(
+      self,
+      interp_weights: FlowFieldVal,
+      interp_weights_neumann: FlowFieldVal,
+      ib_interior_mask: FlowFieldVal,
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:1706]: Remove this variable from this function as it's not used.

File: swirl_lm/boundary_condition/immersed_boundary_method.py [hunk 5]
```diff
@@ -1124,6 +1699,252 @@ def _apply_feedback_force_1d_interp(
...
+          dtype=tf.float32
+      )
+
+      return tf.nest.map_structure(
+          lambda _u, _u_F: -beta * (_u - _u_F), u, gp_tensor
+      )
+
+    return get_ib_force
+
+  def _apply_ghost_point_method(
```
Comment (john-qingwang) [inline swirl_lm/boundary_condition/immersed_boundary_method.py:1853]: I think we should be able to consolidate this function with the other direct forcing methods by replacing the force function.
